cpp-client/Dockerfile

```diff
@@ -26,9 +26,38 @@ RUN set -eux; \
 #
 # Download and build dependent libraries.
 #
+# Be mindful about the order below:
+# 1. Some libraries have cross dependencies (eg, arrow depends on grpc).
+# 2. If something fails, the docker cache at that point gets invalidated;
+#    so it makes sense to put more brittle stuff later on.
+#
```

Comment:
I'm not sure it's worth the added complexity to break these all out as separate RUNs. If something does fail, you'll likely need to make a change to `build-dependencies.sh`, in which case all of these steps will be invalidated anyways. If we had `build-protobuf.sh`, `build-re2.sh`, etc, then re-building just one part of it might save some time.

From a dev perspective though, I think the changes to build-dependencies.sh are useful - just not sure about actually using it here in Dockerfile.